R/data.table.R

```diff
@@ -2457,30 +2458,32 @@ setnames = function(x,old,new,skip_absent=FALSE) {
     if (is.numeric(old)) {
       if (length(sgn <- unique(sign(old))) != 1L)
         stop("Items of 'old' is numeric but has both +ve and -ve indices.")
-      tt = abs(old)<1L | abs(old)>length(x) | is.na(old)
-      if (any(tt)) stop("Items of 'old' either NA or outside range [1,",length(x),"]: ",paste(old[tt],collapse=","))
+      tt = abs(old)<1L | abs(old)>n_col | is.na(old)
+      if (any(tt)) stop("Items of 'old' either NA or outside range [1,",n_col,"]: ", brackify(old[tt]))
       i = if (sgn == 1L) as.integer(old) else seq_along(x)[as.integer(old)]
-      if (any(duplicated(i))) stop("Some duplicates exist in 'old': ",paste(i[duplicated(i)],collapse=","))
+      if (any(idx <- duplicated(i))) stop("Some duplicates exist in 'old': ", brackify(i[idx]))
     } else {
       if (!is.character(old)) stop("'old' is type ",typeof(old)," but should be integer, double or character")
-      if (any(duplicated(old))) stop("Some duplicates exist in 'old': ", paste(old[duplicated(old)],collapse=","))
-      i = chmatch(old,names(x))
+      if (any(idx <- duplicated(old))) stop("Some duplicates exist in 'old': ", brackify(old[idx]))
+      i = chmatch(old, names_x)
       if (anyNA(i)) {
         if (isTRUE(skip_absent)) {
-          w = old %chin% names(x)
+          w = old %chin% names_x
           old = old[w]
           new = new[w]
           i = i[w]
         } else {
-          stop("Items of 'old' not found in column names: ",paste(old[is.na(i)],collapse=","), ". Consider skip_absent=TRUE.")
+          stop("Items of 'old' not found in column names: ", brackify(old[is.na(i)]), ". Consider skip_absent=TRUE.")
         }
       }
-      if (any(tt<-!is.na(chmatch(old,names(x)[-i])))) stop("Some items of 'old' are duplicated (ambiguous) in column names: ",paste(old[tt],collapse=","))
+      # don't bother going further if there's nothing to change, #3783
+      if (identical(names_x[i], new)) return(invisible(x))
```

Comment:
is there a chance that either `new` or `old` are named vectors? if they are we could use `identical` using `unname` on both, as names will not be relevant, and `identical` is sensitive to names.

Reply:
Thought of that, but figured, pretty low payoff -- this is a (small) efficiency improvement & named objects are slightly less efficient in the first place.

Reply:
I mean that this line could be made into `identical(unname(names_x[i]), unname(new))`, unless it is already known that names were stripped out before reaching this `if`. Otherwise `setnames(dt, c(irrelevant="column_a"), c(irrelevant2="column_a")` will not be escaped while it could.